### PR TITLE
frontend dataset view

### PR DESCRIPTION
**Pull request by seve**
#675 

https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/675

Adds dataset view functionality to the collection view
Shows upload status and dataset metadata when available

#### Reviewers
**Functional:** 
@tihuan 
**Readability:** 
@MDunitz 

NOTES:
1. Please note that the checkbox will be replaced with radio button when we implement download files in [Part 3](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/705undefined)
2. The Download button is not functional until [Part 3](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/705undefined) is complete too

---

## Changes
* `frontend/src/common/API.ts`
  * Add status endpoint
 * `frontend/src/common/entities.ts`
   * Add entities for dataset upload status 
 * `frontend/src/common/queries/datasets.ts` + `frontend/src/common/queries/entities.ts`
    * Add query for dataset status
  * `frontend/src/components/Collections/components/Grid/components/DatasetsGrid/index.tsx`
    * create grid for viewing datasets 
  * `frontend/src/components/Collections/components/Grid/components/Row/DatasetRow/index.tsx`
    * Dataset row component, fetches metadata and upload status
  * `frontend/src/components/DropboxChooser/index.tsx` + `frontend/src/views/Collection/index.tsx`
    * Include filename in dropbox chooser
   * `frontend/src/views/Collection/components/DatasetUploadToast/index.tsx`
     * Create toast for communicating dataset upload
    
---

### QA Steps

1. Navigate to a private collection and Follow CTA to upload dataset
   1. Dataset should appear on grid and show uploading status
   1. It will take several minutes for any progress to be shown 
1. At successful upload a toast will appear
   1. Validation process indicator would be shown now, but backend is not currently communicating this
1. After validation and conversion completes, explore button and metadata should populate

1. Erroneous datasets will error and communicate error state via icon in title column
   1. Validation errors are currently communicated as upload error due to backend 

---
image

image

image

### CODE REVIEW

**Comment by tihuan** [comment]
Hi @hthomas-czi quick question on this checkbox!

Screen Shot 2021-01-05 at 12 28 53 PM

Screen Shot 2021-01-05 at 12 30 45 PM


Is it intended to allow the users to select multiple datasets before clicking on "Download" or should it be radio button, so we restrict the user to only select one dataset at a time? Because the Download dropbox seems to be designed to only display one dataset's files?

Thank you!

CC: @seve

**Comment by seve** [comment]
Design updates were made yesterday, will be updating the branch today

**Comment by tihuan** [comment]
Added @hthomas-czi for UX approval! Thank you!

File: frontend/src/common/queries/datasets.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+): Promise<DatasetUploadStatus> {
+  const url = apiTemplateToUrl(API_URL + API.DATASET_STATUS, { dataset_uuid });
+  return await (await fetch(url, DEFAULT_FETCH_OPTIONS)).json();
+}
+
+export function useDatasetStatus(dataset_uuid: string, shouldFetch: boolean) {
+  return useQuery<DatasetUploadStatus>(
+    [USE_DATASET_STATUS, dataset_uuid],
+    fetchDatasetStatus,
+    { enabled: shouldFetch, refetchInterval: 10 * 1000 }
```
Comment (MDunitz) [inline frontend/src/common/queries/datasets.ts:26]: could you make the refetchInterval a constant `TEN_SECONDS`

File: frontend/src/components/DropboxChooser/index.tsx [hunk 2]
```diff
@@ -17,20 +18,23 @@ const AsyncContent = loadable(
...
-};
+  name: string;
+}
+
+export type UploadingFile = DropboxFile & Partial<Dataset>;
 
 const DROPBOX_OPTIONS = {
   extensions: [".h5ad"],
-  sizeLimit: 30 * 10 ** 30, // 30GB
+  sizeLimit: 30 * 2 ** 30, // 30GB
```
Comment (MDunitz) [inline frontend/src/components/DropboxChooser/index.tsx:30]: Could this be a constant?

**Review by hthomas-czi — COMMENTED** [review]
@seve @tihuan I see the dataset that's being uploaded appears in the middle of other datasets (some succeeded, other failed). Datasets that should appear sequentially based on when they were uploaded, so I would expect datasets that are being uploaded to appear at the bottom of the table. How are they being sorted currently?

File: frontend/src/components/Collections/components/Grid/components/Row/DatasetRow/index.tsx [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    datasetStatus.conversion_cxg_status === CONVERSION_STATUS.CONVERTING ||
+    datasetStatus.conversion_rds_status === CONVERSION_STATUS.CONVERTING ||
+    datasetStatus.conversion_loom_status === CONVERSION_STATUS.CONVERTING
+  )
+    return true;
+
+  return false;
+};
+
+// const checkIfComplete = (datasetStatus: DatasetUploadStatus): boolean => {
```
Comment (MDunitz) [inline frontend/src/components/Collections/components/Grid/components/Row/DatasetRow/index.tsx:84]: remove?

**Review by MDunitz — COMMENTED** [review]
some small nits around constants and I dont see any tests but otherwise lgtm

**Comment by tihuan** [comment]
Thanks so much for your super detailed review, @MDunitz ! To avoid merge conflicts, I will make your suggested changes in this [PR](https://github.com/chanzuckerberg/corpora-data-portal/pull/809) 🙏 Thank you!!

File: frontend/src/components/Collections/components/Grid/components/Row/DatasetRow/index.tsx (hunk 3 above)
Comment (tihuan) [reply frontend/src/components/Collections/components/Grid/components/Row/DatasetRow/index.tsx:84]: Yes! Done 🥇

File: frontend/src/components/DropboxChooser/index.tsx (hunk 2 above)
Comment (tihuan) [reply frontend/src/components/DropboxChooser/index.tsx:30]: Great idea! Done!

Changed to:

```ts
const UPLOAD_SIZE_LIMIT_BYTES = 30 * 2 ** 30; // 30GB

const DROPBOX_OPTIONS = {
  extensions: [".h5ad"],
  sizeLimit: UPLOAD_SIZE_LIMIT_BYTES,
};
```

File: frontend/src/common/queries/datasets.ts (hunk 1 above)
Comment (tihuan) [reply frontend/src/common/queries/datasets.ts:26]: Yes! It's now:

```ts
const REFETCH_INTERVAL_MS = 10 * 1000;

export function useDatasetStatus(dataset_uuid: string, shouldFetch: boolean) {
  return useQuery<DatasetUploadStatus>(
    [USE_DATASET_STATUS, dataset_uuid],
    fetchDatasetStatus,
    { enabled: shouldFetch, refetchInterval: REFETCH_INTERVAL_MS }
  );
}
```

**Comment by tihuan** [comment]
> @seve @tihuan I see the dataset that's being uploaded appears in the middle of other datasets (some succeeded, other failed). Datasets that should appear sequentially based on when they were uploaded, so I would expect datasets that are being uploaded to appear at the bottom of the table. How are they being sorted currently?

I had the same question before, and what I observed is that FE is just returning the datasets in the order supplied by the backend. 

FE can sort the datasets from oldest to newest for sure 👍 Will make that change in this [PR](https://github.com/chanzuckerberg/corpora-data-portal/pull/809)! Thank you!

**Review by tihuan — APPROVED** [review]
LGTM! Pending @hthomas-czi 's last review, since @MDunitz 's feedback has been addressed 👍 😄 

Thanks all!

**Pull request merged**